src/MSBuildBinLogQuery/Component/Component.cs

```diff
@@ -0,0 +1,67 @@
+using System;
+using System.Collections.Generic;
+using Microsoft.Build.Logging.Query.Messaging;
+
+namespace Microsoft.Build.Logging.Query.Component
+{
+    public abstract class Component
+    {
+        public abstract Component Parent { get; }
+        public IReadOnlyList<Message> Messages => _messages;
+        public IReadOnlyList<Warning> Warnings => _warnings;
+        public IReadOnlyList<Error> Errors => _errors;
+        
+        /// <summary>
+        /// `AllMessages`, `AllWarnings` and `AllErrors` are messages, warnings
+        /// and errors that occur under this component, either directly or
+        /// indirectly.
+        /// </summary>
+        public IReadOnlyList<Message> AllMessages => _allMessages;
+        public IReadOnlyList<Warning> AllWarnings => _allWarnings;
+        public IReadOnlyList<Error> AllErrors => _allErrors;
+
+        private readonly List<Message> _messages;
+        private readonly List<Warning> _warnings;
+        private readonly List<Error> _errors;
+        private readonly List<Message> _allMessages;
+        private readonly List<Warning> _allWarnings;
+        private readonly List<Error> _allErrors;
+
+        public Component()
+        {
+            _messages = new List<Message>();
+            _warnings = new List<Warning>();
+            _errors = new List<Error>();
+
+            _allMessages = new List<Message>();
+            _allWarnings = new List<Warning>();
+            _allErrors = new List<Error>();
+        }
+
+        public void AddMessage(Message message)
+        {
+            _messages.Add(message);
+            GoUpToRoot(component => component._allMessages.Add(message));
+        }
+
+        public void AddWarning(Warning warning)
+        {
+            _warnings.Add(warning);
+            GoUpToRoot(component => component._allWarnings.Add(warning));
+        }
+
+        public void AddError(Error error)
+        {
+            _errors.Add(error);
+            GoUpToRoot(component => component._allErrors.Add(error));
```

Comment:
This is fine for now but in the future it may be better to compute this top-down via child relationships instead of storing it bottom-up as here.

Reply:
Do you mean computing `AllMessages` and other stuff by a depth-first search recursively? I considered it, but that will have a worse time complexity. The bottom-up approach is O(1), but the top-down would be O(n).

Reply:
True. I don't know how often we will want to do this operation specifically, so we can decide on the memory/time tradeoff later. I suspect that we'll wind up doing so many other graph-walking operations that this won't seem worth treating specially.

Reply:
Got it. We can refactor this later.